src/math/bvh.js

```diff
@@ -0,0 +1,456 @@
+import { Vec3 } from '../math/vec3.js';
+import { BoundingBox } from '../shape/bounding-box.js';
+
+/** @typedef {import('../shape/ray.js').Ray} Ray */
+/** @typedef {import('../math/tri.js').Tri} Tri */
+
+class BVHNode {
+    /**
+     * Create a new BVHNode
+     *
+     * @param {Vec3} [aabbMin] - The bounds of the BVH node
+     * @param {Vec3} [aabbMax] - The bounds of the BVH node
+     * @param {number} [leftFirst] - The index of the left child of the BVH node
+     * @param {number} [triCount] - The number of primitives
+     * @ignore
+     */
+    constructor(aabbMin, aabbMax, leftFirst, triCount) {
+        this.aabbMin = aabbMin || new Vec3();
+        this.aabbMax = aabbMax || new Vec3();
+        this.leftFirst = leftFirst || null;
+        this.triCount = triCount || 0;
+    }
+
+    /**
+     *
+     * @returns {boolean} whether the BVHNode is a leaf node
+     * @ignore
+     */
+    isLeaf() {
+        return this.triCount > 0;
+    }
+}
+
+
+/**
+ * Bin used for BVH Tree construction
+ *
+ * @ignore
+ */
+class Bin {
+    constructor(BINS) {
+        this.BINS = BINS;
+        this.bounds = new BoundingBox(new Vec3(), new Vec3());
+        this.triCount = 0;
+    }
+}
+
+/**
+ * Used to store the BVH, triangles and minimum ray intersection distance for a single mesh
+ */
+class BVHGlobal {
```

Comment:
usually we try the filename to match the main class the file exports. I'd probably rename the class to be Bvh.

Reply:
Renamed to Bvh